### PR TITLE
[SPARK-34944][SQL][TESTS] Replace bigint with int for web_returns and store_returns in TPCDS tests to employ correct data type 

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

According to  http://www.tpc.org/tpc_documents_current_versions/pdf/tpc-ds_v2.9.0.pdf

```
 2.2.2 Datatype
2.2.2.1 Each column employs one of the following datatypes:
a) Identifier means that the column shall be able to hold any key value generated for that column.
b) Integer means that the column shall be able to exactly represent integer values (i.e., values in increments of
1) in the range of at least ( − 2n − 1) to (2n − 1 − 1), where n is 64.
c) Decimal(d, f) means that the column shall be able to represent decimal values up to and including d digits,
of which f shall occur to the right of the decimal place; the values can be either represented exactly or
interpreted to be in this range.
d) Char(N) means that the column shall be able to hold any string of characters of a fixed length of N.
Comment: If the string that a column of datatype char(N) holds is shorter than N characters, then trailing
spaces shall be stored in the database or the database shall automatically pad with spaces upon retrieval such
that a CHAR_LENGTH() function will return N.
e) Varchar(N) means that the column shall be able to hold any string of characters of a variable length with a
maximum length of N. Columns defined as "varchar(N)" may optionally be implemented as "char(N)".
f) Date means that the column shall be able to express any calendar day between January 1, 1900 and
December 31, 2199.
2.2.2.2 The datatypes do not correspond to any specific SQL-standard datatype. The definitions are provided to
highlight the properties that are required for a particular column. The benchmark implementer may employ any internal representation or SQL datatype that meets those requirements.
```

This PR proposes that we use int for identifiers instead of bigint to reach a compromise with TPC-DS Standard Specification.

After this PR, the field schemas are now consistent with those DDLs in the `tpcds.sql` from tpc-ds tool kit, see https://gist.github.com/yaooqinn/b9978a77bbf4f871a95d6a9103019907

### Why are the changes needed?

reach a compromise with TPC-DS Standard Specification

### Does this PR introduce _any_ user-facing change?

no test only

### How was this patch tested?

test only

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @HyukjinKwon please take a look, thanks

**Comment by maropu** [comment]
It seems we need to update the TPCDS query results accordingly if we change the schemas. I'll check this later.

**Comment by yaooqinn** [comment]
> It seems we need to update the TPCDS query results accordingly if we change the schemas. I'll check this later.

Shall we comment out the action temporarily?  I tried to re-generate the result golden file but as the data is static the job failed with unmatched parquet decoding. The static data is not comfy enough if the write-side involving data or schema change itself. Maybe we should generate the data within the test itself, but I don't know if the GA could be able to handle such a large workload.

**Comment by maropu** [comment]
@yaooqinn Yea, I'm planning to do something for the issue in a few weeks. Please see the related discussion: https://github.com/apache/spark/pull/31886#discussion_r603307756

**Comment by HyukjinKwon** [comment]
Thank you @maropu for following up.

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -462,61 +462,61 @@ jobs:
       run: |
         ./build/sbt -Pyarn -Pmesos -Pkubernetes -Phive -Phive-thriftserver -Phadoop-cloud -Pkinesis-asl -Phadoop-2.7 compile test:compile
 
-  tpcds-1g:
-    name: Run TPC-DS queries with SF=1
```
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:466]: Can we simply skip it for now? e.g.) `if: ${{ '1' == '0' }}`

Comment (yaooqinn) [reply .github/workflows/build_and_test.yml:466]: SGTM, thanks~

**Comment by HyukjinKwon** [comment]
I am good with going ahead but I will leave it to @maropu since he's driving this.

**Comment by maropu** [comment]
The citation in the PR description refers to the TPCDS v2.9 doc? http://www.tpc.org/tpc_documents_current_versions/pdf/tpc-ds_v2.9.0.pdf Could you add the link there?

Btw, for better commit logs, could you write what this PR proposes clearly in the description? This PR proposes that we use int for identifeirs instead of bigint, right?

**Comment by yaooqinn** [comment]
thanks @maropu, the suggestion looks good and I've updated the PR title and desc

**Comment by maropu** [comment]
I think the propose itself looks reasonable, so could you add comments somewhere (TPCDSBase) about how we map the TPCDS doc's `Datatype`s (e.g., `identifier` and `integer`) into Spark types?

**Comment by yaooqinn** [comment]
> I think the propose itself looks reasonable, so could you add comments somewhere (TPCDSBase) about how we map the TPCDS doc's `Datatype`s (e.g., `identifier` and `integer`) into Spark types?

```java
/**
 * Base trait for TPC-DS related tests.
 *
 * Datatype mapping for TPC-DS and Spark SQL,
 * see more http://www.tpc.org/tpc_documents_current_versions/pdf/tpc-ds_v2.9.0.pdf
 *
 *    |---------------|---------------|
 *    |    TPC-DS     |  Spark  SQL   |
 *    |---------------|---------------|
 *    |  Identifier   |      INT      |
 *    |---------------|---------------|
 *    |    Integer    |      INT      |
 *    |---------------|---------------|
 *    | Decimal(d, f) | Decimal(d, f) |
 *    |---------------|---------------|
 *    |    Char(N)    |    Char(N)    |
 *    |---------------|---------------|
 *    |  Varchar(N)   |  Varchar(N)   |
 *    |---------------|---------------|
 *    |     Date      |     Date      |
 *    |---------------|---------------|
 *
 * Remarks:
 * The TPC-DS spec requires benchmark implementer may employ any internal representation or SQL
 * datatype that meets the following requirements:
 * 1) Identifier means that the column shall be able to hold any key value generated for that
 *    column.
 * 2) Integer means that the column shall be able to exactly represent integer values (i.e.,
 *    values in increments of 1) in the range of at least −2^(n − 1) to 2^(n − 1) − 1,
 *    where n is 64.
 * 3) Decimal(d, f) means that the column shall be able to represent decimal values up to and
 *    including d digits,of which f shall occur to the right of the decimal place; the values can be
 *    either represented exactly or interpreted to be in this range.
 * 4) Char(N) means that the column shall be able to hold any string of characters of a fixed
 *    length of N.
 * 5) Varchar(N) means that the column shall be able to hold any string of characters of a
 *    variable length with a maximum length of N. Columns defined as "varchar(N)" may optionally
 *    be implemented as "char(N)".
 * 6) Date means that the column shall be able to express any calendar day
 *    between January 1, 1900 and December 31, 2199.
 */
```

Added these comment to the class head

**Comment by yaooqinn** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/TPCDSBase.scala [hunk 2]
```diff
@@ -21,6 +21,49 @@ import org.apache.spark.sql.catalyst.TableIdentifier
...
+ *
+ * Datatype mapping for TPC-DS and Spark SQL, see more at:
+ *   http://www.tpc.org/tpc_documents_current_versions/pdf/tpc-ds_v2.9.0.pdf
+ *
+ *    |---------------|---------------|
+ *    |    TPC-DS     |  Spark  SQL   |
+ *    |---------------|---------------|
+ *    |  Identifier   |      INT      |
+ *    |---------------|---------------|
+ *    |    Integer    |      INT      |
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/TPCDSBase.scala:36]: > One thing might be clear that we should replace bigint type which is now used in web_returns and store_returns with int type.
Another thing that might need to be further discussed is - shall we use bigint for all the integer columns in the TPCDS Data Definition to meet 2.2.2.1 b)?

The statement in the spec below implicitly suggests `Integer` should be bigint?
```
b) Integer means that the column shall be able to exactly represent integer values (i.e., values in increments of
1) in the range of at least ( − 2n − 1) to (2n − 1 − 1), where n is 64.
```

File: sql/core/src/test/scala/org/apache/spark/sql/TPCDSBase.scala [hunk 3]
```diff
@@ -21,6 +21,49 @@ import org.apache.spark.sql.catalyst.TableIdentifier
...
+ *    including d digits,of which f shall occur to the right of the decimal place; the values can be
+ *    either represented exactly or interpreted to be in this range.
+ * 4. Char(N) means that the column shall be able to hold any string of characters of a fixed
+ *    length of N.
+ * 5. Varchar(N) means that the column shall be able to hold any string of characters of a
+ *    variable length with a maximum length of N. Columns defined as "varchar(N)" may optionally
+ *    be implemented as "char(N)".
+ * 6. Date means that the column shall be able to express any calendar day
+ *    between January 1, 1900 and December 31, 2199.
+ *
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/TPCDSBase.scala:65]: nit: remove this blank linke.

File: sql/core/src/test/scala/org/apache/spark/sql/TPCDSBase.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/TPCDSBase.scala:36]: > After this PR, the field schemas are now consistent with those DDLs in the `tpcds.sql` from tpc-ds tool kit, see https://gist.github.com/yaooqinn/b9978a77bbf4f871a95d6a9103019907

removed, according to this line

**Comment by yaooqinn** [comment]
gentle ping @maropu

**Review by cloud-fan — APPROVED** [review]
+1 to be consistent with the table schema in the official tpcds toolkit

File: sql/core/src/test/scala/org/apache/spark/sql/TPCDSBase.scala [hunk 4]
```diff
@@ -21,6 +21,47 @@ import org.apache.spark.sql.catalyst.TableIdentifier
...
+ *    | Decimal(d, f) | Decimal(d, f) |
+ *    |---------------|---------------|
+ *    |    Char(N)    |    Char(N)    |
+ *    |---------------|---------------|
+ *    |  Varchar(N)   |  Varchar(N)   |
+ *    |---------------|---------------|
+ *    |     Date      |     Date      |
+ *    |---------------|---------------|
+ *
+ * Remarks:
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/TPCDSBase.scala:47]: these remarks don't seem to be useful. It's more important to mention that the table schema matches the official tpcds toolkit.

File: .github/workflows/build_and_test.yml [hunk 5]
```diff
@@ -473,6 +473,8 @@ jobs:
     runs-on: ubuntu-20.04
     env:
       SPARK_LOCAL_IP: localhost
+    # TODO: Enable this when the data in 'Cache TPC-DS generated data' fits the new schema
```
Comment (maropu) [inline .github/workflows/build_and_test.yml:476]: Could you use the ref 556111e35d400f56cb0625dc16e9063d54628320 to checkout  `maropu/spark-tpcds-sf-1`? I updated these schemas and re-generated data in https://github.com/maropu/spark-tpcds-sf-1/commit/556111e35d400f56cb0625dc16e9063d54628320

**Comment by cloud-fan** [comment]
There are conflicts...

File: .github/workflows/build_and_test.yml (hunk 5 above)
Comment (yaooqinn) [reply .github/workflows/build_and_test.yml]: i verifed locally, seemed Okay, thanks @maropu. good job!

File: .github/workflows/build_and_test.yml [hunk 6]
```diff
@@ -485,11 +485,10 @@ jobs:
         restore-keys: |
           tpcds-
     - name: Checkout TPC-DS (SF=1) generated data repository
-      if: steps.cache-tpcds-sf-1.outputs.cache-hit != 'true'
```
Comment (maropu) [inline .github/workflows/build_and_test.yml:488]: We need to always fetch the data?

Comment (yaooqinn) [reply .github/workflows/build_and_test.yml:488]: It failed the first time, so I turned it off, shall I turn it on now

**Review by maropu — APPROVED** [review]
Looks fine if the GA tests can pass.

File: .github/workflows/build_and_test.yml (hunk 6 above)
Comment (yaooqinn) [reply .github/workflows/build_and_test.yml]:
```
13:37:51.854 ERROR org.apache.spark.executor.Executor: Exception in task 0.0 in stage 315.0 (TID 204)
java.lang.UnsupportedOperationException: org.apache.parquet.column.values.dictionary.PlainValuesDictionary$PlainLongDictionary
	at org.apache.parquet.column.Dictionary.decodeToInt(Dictionary.java:45)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetDictionary.decodeToInt(ParquetDictionary.java:38)
	at org.apache.spark.sql.execution.vectorized.OnHeapColumnVector.getInt(OnHeapColumnVector.java:298)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:759)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:344)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:898)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:898)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:498)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1437)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:501)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

```
Do we need to disable the cache and merge this PR, then enable the cache again? Still seeing the above error if cache on

Comment (cloud-fan) [reply .github/workflows/build_and_test.yml]: What's the mechanism to refresh the cache?

Comment (yaooqinn) [reply .github/workflows/build_and_test.yml]: I guess it's based on the catch key here https://github.com/apache/spark/pull/32037/files#diff-48c0ee97c53013d18d6bbae44648f7fab9af2e0bf5b0dc1ca761e18ec5c478f2R484,  which is based on the cache data itself(seems a mistake we made before?), and seems that it never changed if we have no opportunity to change the cache data entirely by disabling it. Or let me try to modify the key to rely on the lastest revision of maropu/tpcds-sf-1

Comment (cloud-fan) [reply .github/workflows/build_and_test.yml]: > rely on the lastest revision of maropu/tpcds-sf-1

SGTM

Comment (yaooqinn) [reply .github/workflows/build_and_test.yml:488]: FYI, https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows#matching-a-cache-key

Comment (maropu) [reply .github/workflows/build_and_test.yml:488]: Yea, it looks better to check if it can work correctly without `restore-keys`.

**Comment by yaooqinn** [comment]
All tests pass except a known issue of GA environment` Cause: java.net.BindException: Port in use: fv-az201-716:0` which is not related to this PR.
I am going to merge this, thanks all for reviewing and help

merging to master

**Pull request closed without merging**